session/filesync/filesync.go

```diff
@@ -337,3 +357,44 @@ func (e InvalidSessionError) Error() string {
 func (e InvalidSessionError) Unwrap() error {
 	return e.err
 }
+
+func encodeOpts(opts map[string][]string) map[string][]string {
+	md := make(map[string][]string, len(opts))
+	for k, v := range opts {
+		out := make([]string, len(v))
+		for i, s := range v {
+			out[i] = encodeStringForHeader(s)
+		}
+		md[k] = out
+	}
+	return md
+}
+
+func decodeOpts(opts map[string][]string) map[string][]string {
+	md := make(map[string][]string, len(opts))
+	for k, v := range opts {
+		out := make([]string, len(v))
+		for i, s := range v {
+			out[i], _ = url.QueryUnescape(s)
+		}
+		md[k] = out
+	}
+	return md
+}
+
+// encodeStringForHeader encodes a string value so it can be used in grpc header. This encoding
+// is backwards compatible and avoids encoding ASCII characters.
+func encodeStringForHeader(input string) string {
+	var output strings.Builder
+	for _, runeVal := range input {
+		// Only encode non-ASCII characters.
+		if runeVal > unicode.MaxASCII {
```

Comment:
```golang
		if runeVal > unicode.MaxASCII || runeVal == '%' {
```

would fix this case I think https://github.com/moby/buildkit/pull/3946#issuecomment-1612844955.